### PR TITLE
[Frontend + Runtime] Track allocations

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** [Not all memory allocations can be deallocated using the `--buffer-deallocation` pass.](https://mlir.llvm.org/docs/Bufferization/#buffer-deallocation) Therefore, it is necessary to deallocate them via some other alternative. 

**Description of the Change:** One way to achieve this is to keep track of the calls to memory functions such as malloc and aligned_alloc in a set. When a call to free is made, remove this pointer from the set. At the end of the execution of the function, all remaining allocations in the set should be freed.

**Benefits:** No memory leaks as long as the generated function only calls the generic memory management functions. This is a soft guarantee provided by the memref dialect. It can be broken by other dialects if they choose some lowerings that call other memory management functions.

**Possible Drawbacks:** There is a performance penalty for every insertion and deletion of items in the set. Ideally, these memory allocations would not be needed to be tracked in the set and lack of memory leaks should be guaranteed statically. These wrappers could then still be used to collect statistics without the performance penalty by having some compilation flag like `-DPROFILE_MEM` and having some macro if statements.

Using the following program as a test:

```python
from catalyst import qjit, grad
import pennylane as qml
import jax

device = qml.device("lightning.qubit", wires=1)
@qml.qnode(device, wires=1)
def main(x):
    return 2 * x

@qjit(keep_intermediate=True, verbose=True)
def g(x: jax.core.ShapedArray([], float)):
    h = grad(main)
    return h(x)[0] + 2
```

Running it with inlining disabled, we get a memory leak running with sanitizers in a python-free environment. However after applying these set of patches, with inlining disabled, we get no memory leaks in a python-free environment.

I believe that better and easier testing will be possible once we have a C wrapper generator.

[sc-35382]

### CODE REVIEW

**Comment by dime10** [comment]
Is the cost of insertion into the set really that significant compared to the cost of the allocation itself?

**Comment by erick-xanadu** [comment]
> Is the cost of insertion into the set really that significant compared to the cost of the allocation itself?

@dime10 `std::unordered_set` grows its own buffer. I believe it will allocate twice the capacity it currently has whenever it runs out of memory in its buffer. Using the standard library is not a bad starting point and out of all the data structures in it, I believe that `std::unordered_set` is the best choice. However, I am pointing out that there is indeed a cost associated with maintaining it vs alternatives like statically issuing malloc and free.

There is also the alternative of having an allocation pool, which just keeps allocates and at the end deallocates everything entirely (as opposed to deallocating individual pointers). But this also has its own problem because it means that the peak memory utilization will be higher because the individual allocations will be kept longer in memory.

File: frontend/catalyst/compilation_pipelines.py [hunk 1]
```diff
@@ -326,24 +320,12 @@ def _exec(shared_object_file, func_name, has_return, *args):
...
         retval = CompiledFunction.return_value_ptr_to_numpy(result) if result else None
 
-        if has_return:
-            raw_return = args[0].contents
-            for memref in raw_return:
-                is_constant = memref.allocated == 0xDEADBEEF
-                if is_constant:
-                    continue
-
-                if memref.allocated in numpy_managed_memory:
```
Comment (dime10) [inline frontend/catalyst/compilation_pipelines.py:341]: I think this numpy managed memory is no longer needed right?

File: runtime/lib/capi/RuntimeCAPI.cpp [hunk 2]
```diff
@@ -46,10 +47,58 @@ static auto get_device() -> std::unique_ptr<Catalyst::Runtime::QuantumDevice> &
...
+    std::unordered_set<void *> _impl;
+
+  public:
+    MemoryManager();
+    ~MemoryManager();
+    void insert(void *ptr);
+    void erase(void *ptr);
+};
+
+MemoryManager::MemoryManager() { _impl.reserve(1028); }
```
Comment (dime10) [inline runtime/lib/capi/RuntimeCAPI.cpp:60]: Curious number :D

**Review by dime10 — APPROVED** [review]
Looks great, thanks 💯

File: frontend/catalyst/compilation_pipelines.py (hunk 1 above)
Comment (erick-xanadu) [reply frontend/catalyst/compilation_pipelines.py:341]: You are right. Thanks for pointing it out. https://github.com/PennyLaneAI/catalyst/pull/78/commits/b01a049e99b6c442dd7f659fef3918c6721dadbc

File: runtime/lib/capi/RuntimeCAPI.cpp (hunk 2 above)
Comment (erick-xanadu) [reply runtime/lib/capi/RuntimeCAPI.cpp]: https://github.com/PennyLaneAI/catalyst/pull/78/commits/8e92916e435ceaa406b5e9e3f95f8e547c49b605

**Pull request merged**